pkg/cloudprovider/vsphere/nodemanager.go

```diff
@@ -386,6 +333,203 @@ func (nm *NodeManager) DiscoverNode(nodeID string, searchBy cm.FindVM) error {
+// toIPAddrNetworkNames maps an array of GuestNicInfo to and array of *ipAddrNetworkName.
+func toIPAddrNetworkNames(guestNicInfos []types.GuestNicInfo) []*ipAddrNetworkName {
+	var candidates []*ipAddrNetworkName
+	for _, v := range guestNicInfos {
+		for _, ip := range v.IpAddress {
+			candidates = append(candidates, &ipAddrNetworkName{ipAddr: ip, networkName: v.Network})
+		}
+	}
+	return candidates
+}
+
+// networkNames maps an array of GuestNicInfo to an array of network name strings
+func networkNames(guestNicInfos []types.GuestNicInfo) []string {
+	var existingNetworkNames []string
+	for _, v := range guestNicInfos {
+		existingNetworkNames = append(existingNetworkNames, v.Network)
+	}
+	return existingNetworkNames
+}
+
+// collectMatchesForIPFamily collects all ipAddrNetworkNames that have ips of the
+// desired IP family
+func collectMatchesForIPFamily(ipAddrNetworkNames []*ipAddrNetworkName, ipFamily string) []*ipAddrNetworkName {
+	return filter(ipAddrNetworkNames, func(candidate *ipAddrNetworkName) bool {
+		return matchesFamily(candidate.ip(), ipFamily)
+	})
+}
+
+// matchesFamily detects whether a given IP matches the given IP family.
+func matchesFamily(ip net.IP, ipFamily string) bool {
+	if ipFamily == vcfg.IPv6Family {
+		return ip.To4() == nil
+	}
+
+	if ipFamily == vcfg.IPv4Family {
+		return ip.To4() != nil
+	}
+
+	return false
+}
+
+// filter returns a subset of given ipAddrNetworkNames based on whethe the
+// items in the collection pass the given predicate function.
+func filter(ipAddrNetworkNames []*ipAddrNetworkName, predicate func(*ipAddrNetworkName) bool) []*ipAddrNetworkName {
+	var filtered []*ipAddrNetworkName
+	for _, item := range ipAddrNetworkNames {
+		if predicate(item) {
+			filtered = append(filtered, item)
+		}
+	}
+	return filtered
+}
+
+// findSubnetMatch finds the first *ipAddrNetworkName that has an IP in the
+// given network subnet.
+func findSubnetMatch(ipAddrNetworkNames []*ipAddrNetworkName, networkSubnet *net.IPNet) *ipAddrNetworkName {
+	if networkSubnet != nil {
+		subnetMatches := filter(ipAddrNetworkNames, func(candidate *ipAddrNetworkName) bool {
+			return networkSubnet.Contains(candidate.ip())
+		})
+
+		if len(subnetMatches) > 0 {
+			return subnetMatches[0]
+		}
+		return nil
+	}
+	return nil
+}
+
+// findNetworkNameMatch finds the first *ipAddrNetworkName that matches the
+// given network name, ignoring case.
+func findNetworkNameMatch(ipAddrNetworkNames []*ipAddrNetworkName, networkName string) *ipAddrNetworkName {
+	if networkName != "" {
+		networkNameMatches := filter(ipAddrNetworkNames, func(candidate *ipAddrNetworkName) bool {
+			return strings.EqualFold(networkName, candidate.networkName)
+		})
+
+		if len(networkNameMatches) > 0 {
+			return networkNameMatches[0]
+		}
+		return nil
+	}
+	return nil
+}
+
+// containsCaseInsensitive detects whether a given array of string contains
+// the given string, ignoring case.
+func containsCaseInsensitive(arr []string, str string) bool {
```

Comment:
just a minor one - to me, this function looks like a util functon that we can move to https://github.com/kubernetes/cloud-provider-vsphere/blob/master/pkg/cloudprovider/vsphere/util.go